controller/contour/status.go

```diff
@@ -0,0 +1,164 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contour
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	operatorv1alpha1 "github.com/projectcontour/contour-operator/api/v1alpha1"
+	utilequality "github.com/projectcontour/contour-operator/util/equality"
+	retryable "github.com/projectcontour/contour-operator/util/retryableerror"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+// clock is to enable unit testing
+var clock utilclock.Clock = utilclock.RealClock{}
+
+// syncContourStatus computes the current status of contour and updates status upon
+// any changes since last sync.
+func (r *Reconciler) syncContourStatus(ctx context.Context, contour *operatorv1alpha1.Contour, deployment *appsv1.Deployment, ds *appsv1.DaemonSet) error {
+	var errs []error
+	updated := contour.DeepCopy()
+	updated.Status.AvailableContours = deployment.Status.AvailableReplicas
+	updated.Status.AvailableEnvoys = ds.Status.NumberAvailable
+	updated.Status.Conditions = mergeConditions(updated.Status.Conditions, computeContourAvailableCondition(deployment, ds))
+
+	if utilequality.ContourStatusChanged(contour.Status, updated.Status) {
+		if err := r.Client.Status().Update(ctx, updated); err != nil {
+			errs = append(errs, fmt.Errorf("failed to update contour %s/%s status: %w", contour.Namespace,
+				contour.Name, err))
+		}
+	}
+
+	return retryable.NewMaybeRetryableAggregate(errs)
+}
+
+// computeContourAvailableCondition computes the contour Available status condition
+// type based on deployment and ds. The contour is only available if the deployment's
+// "Available" condition is "True" and the status numberAvailable of ds greater than
+// or equal to 1
+func computeContourAvailableCondition(deployment *appsv1.Deployment, ds *appsv1.DaemonSet) metav1.Condition {
+	if deployment == nil {
```

Comment:
`deployment` cannot be nil or `syncContourStatus` would have panicked on `updated.Status.AvailableContours = deployment.Status.AvailableReplicas`.

(If you do make `syncContourStatus` handle nil values for `deployment` and `ds`, then you could simplify `ensureContour` slightly by declaring the `deploy` and `ds` variables without initializing them to pointers to zero values.)